lib/media/quality_observer.js

```diff
@@ -0,0 +1,276 @@
+   *
+   * @param {!shaka.extern.MediaQualityInfo} mediaQuality
+   * @param {!number} position
+   *  Position in seconds of the quality change.
+   */
+  addMediaQualityChange(mediaQuality, position) {
+    // Remove unneeded QualityChanges before adding the new one
+    this.purgeQualityChangesForContentType_(mediaQuality.contentType);
+    const newQualityChange = {
+      mediaQuality: mediaQuality,
+      position: position,
+    };
+
+    const qualityChanges =
+      this.getContentMediaQuality_(mediaQuality.contentType).qualityChanges;
+    const insertBeforeIndex = qualityChanges.findIndex(
+        (qualityChange) => (qualityChange.position >= position));
+
+    if (insertBeforeIndex >= 0) {
+      const duplicatePositions =
+        (qualityChanges[insertBeforeIndex].position == position) ? 1 : 0;
+      qualityChanges.splice(
+          insertBeforeIndex, duplicatePositions, newQualityChange);
+    } else {
+      qualityChanges.push(newQualityChange);
+    }
+  }
+
+  /**
+   * Determines the media quality at a specific position in the source buffer.
+   *
+   * @param {!number} position
+   *  Position in seconds
+   * @param {!shaka.media.QualityObserver.ContentMediaQuality}
+   *  contentMediaQuality
+   * @return {?shaka.extern.MediaQualityInfo}
+   * @private
+   */
+  static getMediaQualityAtPosition_(position, contentMediaQuality) {
+    // The qualityChanges must be ordered by position ascending
+    // Find the last QualityChange prior to the position
+    const qualityChanges = contentMediaQuality.qualityChanges;
+    for (let i = qualityChanges.length - 1; i >= 0; i--) {
+      const qualityChange = qualityChanges[i];
+      if (qualityChange.position <= position) {
+        return qualityChange.mediaQuality;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Determines if two MediaQualityInfo objects are the same or not.
+   *
+   * @param {?shaka.extern.MediaQualityInfo} mq1
+   * @param {?shaka.extern.MediaQualityInfo} mq2
+   * @return {boolean}
+   * @private
+   */
+  static mediaQualitiesAreTheSame_(mq1, mq2) {
+    if (!mq1 || !mq2) {
+      return false;
+    }
+    if (mq1 === mq2) {
+      return true;
+    }
+    return (mq1.bandwidth == mq2.bandwidth) &&
+      (mq1.audioSamplingRate == mq2.audioSamplingRate) &&
+      (mq1.codecs == mq2.codecs) &&
+      (mq1.contentType == mq2.contentType) &&
+      (mq1.frameRate == mq2.frameRate) &&
+      (mq1.height == mq2.height) &&
+      (mq1.mimeType == mq2.mimeType) &&
+      (mq1.numChannels == mq2.numChannels) &&
+      (mq1.pixelAspectRatio == mq2.pixelAspectRatio) &&
+      (mq1.width == mq2.width);
+  }
+
+  /** @override */
+  poll(positionInSeconds, wasSeeking) {
+    for (const contentMediaQuality of this.contentMediaQualities_.values()) {
+      const mediaQualityAtPosition =
+        shaka.media.QualityObserver.getMediaQualityAtPosition_(
+            positionInSeconds, contentMediaQuality);
+      if (mediaQualityAtPosition &&
```

Comment:
You have this check here because `mediaQualitiesAreTheSame_` considers `null` to not be the same as `null`, right? I think that would be better to move that logic into `mediaQualitiesAreTheSame_`.

Reply:
I am changing `mediaQualitiesAreTheSame_` such that `null` is the same as `null` because `null` != `null` is confusing. I think we need to keep this check.